Add MegaLinter

jtxa opened the pull request:
Adds a GitHub workflow for MegaLinter with initial configuration and some fixed.

Successful linters:
- Bash: bash-exec, shellcheck, shfmt
- Editorconfig
- JSON: eslint-plugin-jsonc, jsonlint, prettier, v8r
- Markdown: markdownlint, markdown-link-check, markdown-table-formatter
- Repository: gitleaks, git_diff, secretlint
- Spelling: cspell, misspell
- YAML: prettier, v8r, yamllint

Switched to be non-blocking:
- CPP: cpplint - out of scope of this PR; changes would need more detailed reviews
- HTML: djlint, htmllint - out of scope of this PR; files in git have troff commands and rest is generated

jtxa commented on .editorconfig:
```diff
@@ -5,13 +5,19 @@ charset = utf-8
 indent_size = 4
 indent_style = space
 insert_final_newline = true
-max_line_length = 80
+max_line_length = 120
 tab_width = 8
 trim_trailing_whitespace = true
 
-[*.yml;*.cmake;CMakeLists.txt]
+[*.{cc,h,h.in}]
+max_line_length = 80
+
+[{*.cmake,CMakeLists.txt}]
 indent_size = 2
 tab_width = 2
 
+[*.{html,man,so}]
+max_line_length = 400
```
I need a decision what the indent size and max line length shall be for each file.
The editorconfigcheck validates if every file is fitting this configuration.
In the past the line length was strictly 80 for every file.

My suggestion:
- consistent indent=4 for every file
- increase max=120, on reviews check if a line is readable/too complex or not
- for C code we might keep 80, please see commit 5e1f054a for what needs to be changed then
- html,man,so: they need a bigger limit today (i don't wanted to touch troff files myself)

sierrafoxtrot replied:
I'd been looking at the line length issue myself also. I'm old enough to almost remember actual terminals in use with 80 col limits and normally I don't like to set this low a limit but it really is in keeping with the original coding style. Given how close we are, I agree, let's just keep 80 with C/C++.

For reasons mentioned above, I'm happy with the 120 elsewhere. Consistent indent of 4 (spaces not tabs) works for me too. 

I'm happy to tweak the troff and soelim files. Once we get your changes merged, I could do a follow-up on those. 120 **should** be workable there. I'll take a look. So for now, agreed, let's keep it large.

jtxa commented on README.md:
In the previous commit I just fixed some issues on the README.
In this commit I rewrote the text completely. I think this compact version better suits a README.
Please say which version to keep.

sierrafoxtrot replied:
I do like yours more. The readme was a last minute thing I added quickly as github "just looked wrong" without it. I briefly tried to generate it from troff but wasn't worth the effort so I copied from the webpage. Like I said, this is much more appealing.

jtxa commented:
@sierrafoxtrot Please have a look at the questions above. If you tell me your preference, I can make this PR ready for merge.

sierrafoxtrot commented:
> Adds a GitHub workflow for MegaLinter with initial configuration and some fixed.
> 
> Successful linters:
> 
>     * JSON: eslint-plugin-jsonc, jsonlint, prettier, v8r
Curious about this one. We don't have any JSON or JS from memory. Does this cover something I'm forgetting?

jtxa commented:
The new cSpell configuration is currently a JSON file.
But I take a look to convert it to YAML, that way comments can also be added and it might look a bit easier.

sierrafoxtrot commented:
> The new cSpell configuration is currently a JSON file. But I take a look to convert it to YAML, that way comments can also be added and it might look a bit easier.

I understand. This work is a huge leap forward Josef so happy to keep it as-is if that is the simplest way forward.

jtxa commented:
A little polishing done here and there since the first push. The major ones are:

- cSpell configuration as YAML: better readability and comments added why some settings were made.
- All YAML files with indent of 2, everything else looks weird because of the arrays using `- `.
- I did check if all the output of the doc scripts is the same as before, and now they are (see #21).
- 
From my side this PR is ready now.

jtxa commented:
Rebase to solve conflicts.
Also deleted shell/awk/ignore files, obsolete after #22

sierrafoxtrot commented on test/00/t0003a.sh:
```diff
@@ -38,7 +38,10 @@ fubar
 
 srec_cat test.in -intel -o test.out -motorola -header HDR > LOG 2>&1
-if test $? -ne 0; then cat LOG; fail; fi
+if test $? -ne 0; then
```
Curious about why proposing a change to multi-line here while leaving line 38 unchanged. I prefer the single line for this kind of boilerplate but happy to be convinced otherwise. (more instances below).

jtxa replied:
This was automatically done by `shfmt`. Couldn't find any hint in the [Google ShellGuide](https://google.github.io/styleguide/shellguide.html).

But I assume the rationale is, that multi-line shall be used if more than 1 command is used.

sierrafoxtrot replied:
That makes sense. I can't find any way to configure this in `shfmt` either. On balance, I really like being able to get consistent formatting so let's just keep the change as is.

sierrafoxtrot reviewed (commented):
Thanks Josef, this is a huge improvement and significantly cleans things up. One query above but I'm keen to get this merged.

marcows commented on doc/CMakeLists.txt:
```diff
@@ -61,52 +43,72 @@ file(GLOB_RECURSE NEW_CHANGE_NOTES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "etc/new
+            # Order of preprocessors may be important
+            -t                     # Preprocess with tbl
+            -s                     # Preprocess with soelim, with search paths:
+            -I ${DocSourcePath}
```
Adding new and moving existing options shouldn't be mixed with style changes.

jtxa replied:
Yes that's true. I tried to kept the numbers of commits low for Scott ;-)

marcows commented on doc/CMakeLists.txt:
```diff
@@ -118,61 +120,103 @@ endif()
+            -man                   # Macro Package man
+            # Order of preprocessors may be important
+            -t                     # Preprocess with tbl
+            -s                     # Preprocess with soelim, with search paths:
```
You write the order may be important, but swap it here in this commit which is mostly about style changes.

jtxa replied:
I don't know groff, and found no hint about this behavior in the options description in the man page.

But checking the output for being identical I recognized that the order of `-P-I${image_stem_name}` had an influence on the filename of the generated png.

jtxa replied:
I did now normalize the order of all groff calls. They order in general seems to be irrelevant, just that one is not.
I left the comment in the code, so the people are aware to be sure to check the output files if those things are changed.

sierrafoxtrot replied:
Yep, the ordering is important except when it isn't :-(. Adding the image stem option was fiddly. Needed to give unique website images but also to support concurrent doc builds (there was a race condition around generating an image for one doc entry being overwritten by another running in parallel). The other main gotchas I encountered were around the args being passed to the preprocesors like soelim and tbl.

jtxa commented:
- Rebased
- Doc-CMake changes splitted to mulitple commits with revised content and messages
- I reverted some of my `ci.yml` changes. There will be a follow-up PR with Windows and MacOS build added.
- The rest is the same as before. I repeated my check that the generated documentation is still the same.
